### DISCUSSION TITLE
Would state charts be a good fit to define declarative business logic in forms?

### DISCUSSION THREAD

**Discussion by manuschillerdev** [statelyai/xstate#2775] [Q&A]
Hi!

We often do client projects for insurance companies, which include forms tons of dependent fields and business logic in them.

* "Show field X only if option A was selected in Dropdown Y."
* "disable field X if the value of input Y is smaller than 2000."
* "use this validation rule for field X only if the user belongs to a special user group."

...and so on.

Currently, I am looking for a *declarative way* and *serializable* way to represent and evaluate these business logic rules across different systems in a microservice architecture. Therefore I would love to have some standard instead of implementing our own solution for that.

Are there any examples of business logic within forms defined in XState? Would XState be a good fit for this scenario? 

I found a lot of examples expressing form "meta state" (`state === "submitting", state==="submitted"), but zero to none expressing business logic within the form in a serializable way.

Thanks a lot for any input on this topic! :)

**Comment by davidkpiano** [comment]
Hey Manu!

This reminds me of the very reason why I created XState in the first place - complex forms with lots of business logic.

Let's take the first example:

> Show field X only if option A was selected in Dropdown Y.

State machines dictate what can and cannot _happen_ in a given state. So, rephrasing this, what I feel is more accurate is:

> Field X can only be changed if option A was selected in dropdown Y.

The difference here is that there is no "showing" or "disabled", etc. logic here - it's just business logic, and whether it is shown/disabled/etc. based on that business logic is the concern of view logic.

So in XState, we can specify something like:

```js
on: {
  'fieldX.change': {
    // only allow this change if some condition is true
    cond: ctx => ctx.fieldY === optionA,
    actions: 'updateFieldX'
  }
}
```

This is actually more robust than just having logic for showing/hiding, because it makes it _impossible_ for the Field X value to change (according to this logic), regardless of whether the field is shown/hidden/disabled/etc.

But then, how do you control the view logic? Given this business logic, we can represent the view logic succinctly as:

> Show field X **only if field X can be changed**

Where that "only if field X can be changed" logic already exists as "only if option A was selected in dropdown Y".

You can then use `state.can(...)`:

```jsx
{state.can({ type: 'fieldX.change', value: ...  }) && <FieldX />}
```

---

I just noticed that there's somewhere we can improve - currently `state.can(...)` is trying to determine if a state can be changed on an exact event, but not on _any_ event with a certain type 🤔

**Reply by manuschillerdev** [reply]
Interesting, thanks!

Could we provide `cond` in a serializable, declarative way as well?
The goal would be to have all business logic in schema files separated from the rendering logic.

Right now, I am experimenting with (ab)using JSON-Schema errors for this kind of logic:
https://stackblitz.com/edit/react-ts-mxxqzd?file=schema.ts

I know, exceptions for control flow - but hey, it works well for suspense 🤷‍♂️ 

The problem is that this approach is obviously limited (e.g. how to distinguish disabled/enabled state from visibility state?)

Maybe we could use subschemas for `cond` and evaluate them generically with a JSON-validator?
Or is there any better way to declaratively define these kinds of conditions with XState?

**Reply by manuschillerdev** [reply]
had a minor issue in the updater logic - sandbox should be fixed now :)
